### PR TITLE
1978 update lfric infrastructure and psydata

### PR DESCRIPTION
**Pull request by hiker**
This adds support for 32- and 64-bit fields to all PSyData libraries. Note that atm we don't have an easy way of testing these libraries (see #1979 for the missing bit), but I have successfully used the LFRic NetCDF extraction library with gungho, which is mixed precision.

This PR:
- adds the new infrastructure files to support 32- and 64-bit fields
- updates all PSyData libraries using jinja to fully support mixed precision.

### CODE REVIEW

**Comment by hiker** [comment]
The CI failure is not my fault, like like python 3.6. is not available anymore??

I have also fixed #1529 with this (i.e. I've also added compilation for eg19).

**Review by TeranIvy — CHANGES_REQUESTED** [review]
Thanks for the updates to the infrastructure and libraries, especially the first one was long overdue.

I also like that infrastructure interface libraries are now Jinja templates as this is much more flexible. They would benefit from some tidying up, though (see the inline comments).

There was nothing to check for `pycodestyle` and `pylint`. I checked building of documentation to be on the safe side and that works.

`pytest` without compilation and `make` in `examples` without compilation also work.

However, things break when I try to compile.

* `pytest --compile` with GCC 11.2.0 returns this:

```
E           cd field; gfortran -Wall -g -fcheck=bound -I /var/tmp/pytest-of-ikavcic/pytest-47/dynamo_wrapper0/configuration -I /var/tmp/pytest-of-ikavcic/pytest-47/dynamo_wrapper0/field -I /var/tmp/pytest-of-ikavcic/pytest-47/dynamo_wrapper0/function_space -I /var/tmp/pytest-of-ikavcic/pytest-47/dynamo_wrapper0/initialisation -I /var/tmp/pytest-of-ikavcic/pytest-47/dynamo_wrapper0/io -I /var/tmp/pytest-of-ikavcic/pytest-47/dynamo_wrapper0/mesh -I /var/tmp/pytest-of-ikavcic/pytest-47/dynamo_wrapper0/operator -I /var/tmp/pytest-of-ikavcic/pytest-47/dynamo_wrapper0/scalar -I /var/tmp/pytest-of-ikavcic/pytest-47/dynamo_wrapper0/utilities -c /net/home/h06/ikavcic/PSyclone_work/PSyclone/src/psyclone/tests/test_files/dynamo0p3/infrastructure//field/field_mod.F90 -o field_mod.o
E           $HOME/PSyclone_work/PSyclone/src/psyclone/tests/test_files/dynamo0p3/infrastructure//field/field_mod.F90:25:7:
E           
E              25 |   use field_r64_mod, only: field_type         => field_r64_type, &
E                 |       1
E           Fatal Error: Cannot open module file 'field_r64_mod.mod' for reading at (1): No such file or directory
E           compilation terminated.
E           make: *** [field/field_mod.o] Error 1
```
* `make compile` in `examples/lfric` also breaks. Trying to run `make clean` and recompiling is not working. The already compiled libraries and modules are not removed, I suspect due to a double slash in paths, e.g.
```
../../../src/psyclone/tests/test_files/dynamo0p3/infrastructure//mesh -I ../../../src/psyclone/tests/test_files/dynamo0p3/infrastructure//operator -I ../../../src/psyclone/tests/test_files/dynamo0p3/infrastructure//scalar -I ../../../src/psyclone/tests/test_files/dynamo0p3/infrastructure//utilities -c scaled_matrix_vector_kernel_mod.F90
```
* `git ignore` should be updated for the `lfric/eg11` example as `git status` after make returns
```
eg11/helmholtz_solver_alg_mod.f90
eg11/helmholtz_solver_alg_psy.f90
```

* `git status` in all examples after `make` returns:
 ```
 gocean/eg3/opencl_kernels_0.cl
 lfric/eg11/helmholtz_solver_alg_mod.f90
lfric/eg11/helmholtz_solver_alg_psy.f90
nemo/eg3/traldf_iso.F90
psyad/eg1/test_harness.f90
psyad/eg1/testkernadj_mod.f90
psyad/eg2/adj_hydrostatic_kernel_mod.F90
psyad/eg2/adj_hydrostatic_kernel_mod.x90
psyad/eg2/adjoint_test_mod.F90
psyad/eg2/adjoint_test_mod.x90
psyad/eg2/psy.f90
```

File: examples/lfric/eg19/Makefile [hunk 1]
```diff
@@ -35,7 +35,20 @@
 
 include ../../common.mk
 
-GENERATED_FILES = psy.f90 alg.f90
+GENERATED_FILES = psy.f90 alg.f90 *.o *.mod
```
[the remark this answers is not in the document]
Comment (hiker) [reply examples/lfric/eg19/Makefile:40]: Done.

File: lib/nan_test/lfric/nan_test.jinja [hunk 2]
```diff
@@ -53,93 +65,152 @@ module nan_test_psy_data_mod
...
     !! @param[in] value The value of the variable.
-    subroutine ProvideField(this, name, value)
+    subroutine DeclareFieldVector_r{{prec}}(this, name, value)
 
-        use field_mod, only : field_type, field_proxy_type
+        implicit none
+
+        class(nan_test_PSyDataType), intent(inout), target :: this
+        character(*), intent(in)                           :: name
+        type(field_r{{prec}}_type), dimension(:), intent(in)     :: value
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/nan_test/lfric/nan_test.jinja:165]: See below.

File: lib/nan_test/lfric/nan_test.jinja [hunk 3]
```diff
@@ -53,93 +65,152 @@ module nan_test_psy_data_mod
...
+    !! @param[in] name The name of the variable (string).
+    !! @param[in] value The vector of fields.
+    subroutine ProvideFieldVector_r{{prec}}(this, name, value)
 
         implicit none
 
         class(nan_test_PSyDataType), intent(inout), target :: this
         character(*), intent(in)                           :: name
-        type(field_type), intent(in)                       :: value
+        type(field_r{{prec}}_type), dimension(:), intent(in)     :: value
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/nan_test/lfric/nan_test.jinja:181]: See below.

File: lib/read_only/lfric/read_only.jinja [hunk 4]
```diff
@@ -58,71 +71,66 @@ module read_only_verify_psy_data_mod
...
 
         ! Declare generic interface for PreDeclareVariable:
-        generic, public :: PreDeclareVariable => &
-                           DeclareField,         &
-                           DeclareFieldVector,   &
-                           DeclareIntField,      &
+        generic, public :: PreDeclareVariable =>  &
+        {% for prec in ALL_PREC %}
+                           DeclareField_r{{prec}},      &
+                           DeclareFieldVector_r{{prec}},&
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/read_only/lfric/read_only.jinja:95]: See below.

File: lib/read_only/lfric/read_only.jinja [hunk 5]
```diff
@@ -58,71 +71,66 @@ module read_only_verify_psy_data_mod
...
         !! the checksum (before a kernel), or compares a checksum (after
         !! a kernel call).
-        generic, public :: ProvideVariable =>  &
-                           ProvideField,       &
-                           ProvideFieldVector, &
-                           ProvideIntField,    &
+        generic, public :: ProvideVariable =>     &
+        {% for prec in ALL_PREC %}
+                           ProvideField_r{{prec}},      &
+                           ProvideFieldVector_r{{prec}},&
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/read_only/lfric/read_only.jinja:107]: See below.

File: lib/read_only/lfric/read_only.jinja [hunk 6]
```diff
@@ -168,7 +175,79 @@ subroutine ProvideField(this, name, value)
...
+    !! @param[in,out] this The instance of the read_only_verify_PSyDataType.
+    !! @param[in] name The name of the variable (string).
+    !! @param[in] value The vector of fields.
+    subroutine ProvideFieldVector_r{{prec}}(this, name, value)
+
+        implicit none
+
+        class(read_only_verify_PSyDataType), intent(inout), target :: this
+        character(*), intent(in)                                   :: name
+        type(field_r{{prec}}_type), dimension(:), intent(in)             :: value
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/read_only/lfric/read_only.jinja:213]: See below.

**Review by hiker — COMMENTED** [review]
You commented in several places about the alignment. FWIW, the alignment is all done so that the produced .f90 files are properly aligned. I decided to do that originally so the produced Fortran code does not look auto-produced.

But I totally get that we could focus on readability of the .jinja files. I don't mind, I just thought I point the reason for this strange indentation out, and leave it to you to decide which is the better solution, since I can't decide ;)

File: lib/extract/netcdf/lfric/kernel_data_netcdf.jinja [hunk 7]
```diff
@@ -1,7 +1,14 @@
+{# Added this as jinja code so that it is understood that the
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/netcdf/lfric/kernel_data_netcdf.jinja]: Done ... here and everywhere.

File: lib/extract/netcdf/lfric/kernel_data_netcdf.jinja [hunk 8]
```diff
@@ -1,7 +1,14 @@
...
+   comment does not apply to THIS file. #}
+{{ "! ================================================== !" }}
+{{ "! THIS FILE IS CREATED FROM THE JINJA TEMPLATE FILE  !" }}
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/netcdf/lfric/kernel_data_netcdf.jinja]: Done ... here and everywhere.

File: lib/extract/netcdf/lfric/kernel_data_netcdf.jinja [hunk 9]
```diff
@@ -1,7 +1,14 @@
...
+{{ "! DO NOT MODIFY DIRECTLY                             !" }}
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/netcdf/lfric/kernel_data_netcdf.jinja]: Done ... here and everywhere.

File: lib/extract/netcdf/lfric/kernel_data_netcdf.jinja [hunk 10]
```diff
@@ -56,120 +65,179 @@ module extract_psy_data_mod
...
 
-    end subroutine DeclareIntField
+    end subroutine WriteField_r{{prec}}
 
     ! -------------------------------------------------------------------------
-    !> This subroutine writes the values of an LFRic real-valued field.
-    !! to the NetCDF file. It uses the corresponding function
-    !! provided by the base class.
-    !! @param[in,out] this The instance of the extract_PsyDataType.
+    !> This subroutine declares an LFric real-valued {{prec}}-bit field vector.
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/netcdf/lfric/kernel_data_netcdf.jinja]: Done.

File: lib/extract/netcdf/lfric/kernel_data_netcdf.jinja [hunk 11]
```diff
@@ -56,120 +65,179 @@ module extract_psy_data_mod
...
+        type(field_r{{prec}}_type), dimension(:), intent(in)        :: value
 
-        type(field_proxy_type) :: value_proxy
+        integer                :: i
+        type(field_r{{prec}}_proxy_type) :: value_proxy
+        character(9)           :: number
 
-        value_proxy = value%get_proxy()
-        call this%ProvideVariable(name, value_proxy%data)
+        ! Provide each component of the vector as an individual 1d array.
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/netcdf/lfric/kernel_data_netcdf.jinja]: Done ... here and everywhere.

File: lib/extract/netcdf/lfric/kernel_data_netcdf.jinja [hunk 12]
```diff
@@ -56,120 +65,179 @@ module extract_psy_data_mod
...
+        ! The base class will re-allocate internal array sizes if required
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/netcdf/lfric/kernel_data_netcdf.jinja]: Done.

File: lib/extract/netcdf/lfric/kernel_data_netcdf.jinja [hunk 13]
```diff
@@ -56,120 +65,179 @@ module extract_psy_data_mod
...
     end type extract_PSyDataType
 
 contains
 
+
+    {% for prec in ALL_PREC %}
+
     ! -------------------------------------------------------------------------
-    !> This subroutine declares an LFRic field with real-valued data.
+    !> This subroutine declares an LFRic field with real-valued
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/netcdf/lfric/kernel_data_netcdf.jinja]: Done ... here and everywhere.

File: lib/extract/netcdf/lfric/kernel_data_netcdf.jinja [hunk 14]
```diff
@@ -56,120 +65,179 @@ module extract_psy_data_mod
...
 
-    end subroutine DeclareField
+    end subroutine DeclareField_r{{prec}}
 
     ! -------------------------------------------------------------------------
-    !> This subroutine declares an LFRic field with integer-valued data.
-    !! It calls the PreDeclareVariable function provided by the base class
-    !! (depending on the type of the argument, e.g. it might call
-    !! DeclareArray1dInt).
+    !> This subroutine writes the values of an LFRic real-valued {{prec}}-bit field.
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/netcdf/lfric/kernel_data_netcdf.jinja]: Done ... here and everywhere.

File: lib/extract/netcdf/lfric/kernel_data_netcdf.jinja (hunk 10 above)
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/netcdf/lfric/kernel_data_netcdf.jinja]: Done ... here and everywhere.

File: lib/extract/netcdf/lfric/kernel_data_netcdf.jinja [hunk 15]
```diff
@@ -56,120 +65,179 @@ module extract_psy_data_mod
...
+        enddo
 
-    end subroutine WriteField
+    end subroutine DeclareFieldVector_r{{prec}}
 
     ! -------------------------------------------------------------------------
-    !> This subroutine writes the values of an LFRic integer-valued field.
-    !! to the NetCDF file. It uses the corresponding function
-    !! provided by the base class.
+    !> This subroutine writes an LFRic real-valued {{prec}}-bit field vector to
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/netcdf/lfric/kernel_data_netcdf.jinja]: Done ... here and everywhere.

File: lib/extract/netcdf/lfric/kernel_data_netcdf.jinja [hunk 16]
```diff
@@ -56,120 +65,179 @@ module extract_psy_data_mod
...
+
+        class(extract_PSyDataType), intent(inout), target :: this
+        character(*), intent(in)                          :: name
+        type(field_r{{prec}}_type), dimension(:), intent(in)        :: value
+
+        integer                :: i
+        type(field_r{{prec}}_proxy_type) :: value_proxy
+        character(9)           :: number
+
+        ! Provide each dimension of the vector as an individual 1d array.
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/netcdf/lfric/kernel_data_netcdf.jinja]: Done ... here and everywhere.

File: lib/extract/netcdf/lfric/kernel_data_netcdf.jinja [hunk 17]
```diff
@@ -56,120 +65,179 @@ module extract_psy_data_mod
...
+        enddo
+
+    end subroutine WriteFieldVector_r{{prec}}
+
+
+
+    {% endfor %}
+
+    ! -------------------------------------------------------------------------
+    !> This subroutine declares an LFRic field with integer-valued data.
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/netcdf/lfric/kernel_data_netcdf.jinja]: Done ... here and everywhere.

File: lib/extract/netcdf/lfric/kernel_data_netcdf.jinja [hunk 18]
```diff
@@ -183,43 +251,34 @@ subroutine WriteIntField(this, name, value)
...
 
-    end subroutine WriteIntField
+    end subroutine DeclareIntField
 
     ! -------------------------------------------------------------------------
-    !> This subroutine declares an LFric real-valued field vector. Each
-    !! component of the vector is stored as a separate variable, using the
-    !! corresponding array function of the base class.
-    !! @param[in,out] this The instance of the extract_PSyDataType.
+    !> This subroutine writes the values of an LFRic integer-valued field.
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/netcdf/lfric/kernel_data_netcdf.jinja]: Done ... here and everywhere.

File: lib/extract/netcdf/lfric/kernel_data_netcdf.jinja [hunk 19]
```diff
@@ -183,43 +251,34 @@ subroutine WriteIntField(this, name, value)
...
+    subroutine WriteIntField(this, name, value)
 
-        use field_mod, only : field_type, field_proxy_type
+        use integer_field_mod, only : integer_field_type, &
+                                      integer_field_proxy_type
 
         implicit none
 
-        class(extract_PSyDataType), intent(inout), target :: this
+        class(extract_PsyDataType), intent(inout), target :: this
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/netcdf/lfric/kernel_data_netcdf.jinja]: Done.

File: lib/extract/standalone/lfric/kernel_data_standalone.jinja [hunk 20]
```diff
@@ -1,7 +1,14 @@
+{# Added this as jinja code so that it is understood that the
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/standalone/lfric/kernel_data_standalone.jinja]: Done ... here and everywhere.

File: lib/extract/standalone/lfric/kernel_data_standalone.jinja [hunk 21]
```diff
@@ -1,7 +1,14 @@
...
+   comment does not apply to THIS file. #}
+{{ "! ================================================== !" }}
+{{ "! THIS FILE IS CREATED FROM THE JINJA TEMPLATE FILE  !" }}
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/standalone/lfric/kernel_data_standalone.jinja]: Done ... here and everywhere.

File: lib/extract/standalone/lfric/kernel_data_standalone.jinja [hunk 22]
```diff
@@ -1,7 +1,14 @@
...
+{{ "! DO NOT MODIFY DIRECTLY                             !" }}
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/standalone/lfric/kernel_data_standalone.jinja]: Done ... here and everywhere.

File: lib/extract/standalone/lfric/kernel_data_standalone.jinja [hunk 23]
```diff
@@ -55,120 +65,179 @@ module extract_psy_data_mod
...
 
-    end subroutine DeclareField
+    end subroutine DeclareField_r{{prec}}
 
     ! -------------------------------------------------------------------------
-    !> This subroutine declares an LFRic field with integer-valued data.
-    !! It calls the PreDeclareVariable function provided by the base class
-    !! (depending on the type of the argument, e.g. it might call
-    !! DeclareArray1dInt).
+    !> This subroutine writes the values of an LFRic real-valued {{prec}}-bit field.
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/standalone/lfric/kernel_data_standalone.jinja]: Done ... here and everywhere.

File: lib/extract/standalone/lfric/kernel_data_standalone.jinja [hunk 24]
```diff
@@ -55,120 +65,179 @@ module extract_psy_data_mod
...
 
-    end subroutine DeclareIntField
+    end subroutine WriteField_r{{prec}}
 
     ! -------------------------------------------------------------------------
-    !> This subroutine writes the values of an LFRic real-valued field.
-    !! to the binary file. It uses the corresponding function
-    !! provided by the base class.
-    !! @param[in,out] this The instance of the extract_PsyDataType.
+    !> This subroutine declares an LFric real-valued {{prec}}-bit field vector.
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/standalone/lfric/kernel_data_standalone.jinja]: Done ... here and everywhere.

File: lib/extract/standalone/lfric/kernel_data_standalone.jinja [hunk 25]
```diff
@@ -55,120 +65,179 @@ module extract_psy_data_mod
...
 
-        type(field_proxy_type) :: value_proxy
+        integer                :: i
+        type(field_r{{prec}}_proxy_type) :: value_proxy
+        character(9)           :: number
 
-        value_proxy = value%get_proxy()
-        call this%ProvideVariable(name, value_proxy%data)
+        ! Provide each component of the vector as an individual 1d array.
+        ! The base class will re-allocate internal array sizes if required
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/standalone/lfric/kernel_data_standalone.jinja]: Done.

File: lib/extract/standalone/lfric/kernel_data_standalone.jinja [hunk 26]
```diff
@@ -55,120 +65,179 @@ module extract_psy_data_mod
...
+        enddo
 
-    end subroutine WriteField
+    end subroutine DeclareFieldVector_r{{prec}}
 
     ! -------------------------------------------------------------------------
-    !> This subroutine writes the values of an LFRic integer-valued field.
-    !! to the binary file. It uses the corresponding function
-    !! provided by the base class.
+    !> This subroutine writes an LFRic real-valued {{prec}}-bit field vector to
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/standalone/lfric/kernel_data_standalone.jinja]: Done ... here and everywhere.

File: lib/extract/standalone/lfric/kernel_data_standalone.jinja [hunk 27]
```diff
@@ -55,120 +65,179 @@ module extract_psy_data_mod
...
+        enddo
+
+    end subroutine WriteFieldVector_r{{prec}}
+
+
+
+    {% endfor %}
+
+    ! -------------------------------------------------------------------------
+    !> This subroutine declares an LFRic field with integer-valued data.
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/standalone/lfric/kernel_data_standalone.jinja]: Done ... here and everywhere.

File: lib/extract/standalone/lfric/kernel_data_standalone.jinja [hunk 28]
```diff
@@ -182,43 +251,34 @@ subroutine WriteIntField(this, name, value)
...
 
-    end subroutine WriteIntField
+    end subroutine DeclareIntField
 
     ! -------------------------------------------------------------------------
-    !> This subroutine declares an LFric real-valued field vector. Each
-    !! component of the vector is stored as a separate variable, using the
-    !! corresponding array function of the base class.
-    !! @param[in,out] this The instance of the extract_PSyDataType.
+    !> This subroutine writes the values of an LFRic integer-valued field.
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/standalone/lfric/kernel_data_standalone.jinja]: Done ... here and everywhere.

File: lib/extract/standalone/lfric/kernel_data_standalone.jinja [hunk 29]
```diff
@@ -182,43 +251,34 @@ subroutine WriteIntField(this, name, value)
...
+    !! to the binary file. It uses the corresponding function
+    !! provided by the base class.
+    !! @param[in,out] this The instance of the extract_PsyDataType.
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/standalone/lfric/kernel_data_standalone.jinja]: Done.

File: lib/extract/standalone/lfric/kernel_data_standalone.jinja [hunk 30]
```diff
@@ -182,43 +251,34 @@ subroutine WriteIntField(this, name, value)
...
+    subroutine WriteIntField(this, name, value)
 
-        use field_mod, only : field_type, field_proxy_type
+        use integer_field_mod, only : integer_field_type, &
+                                      integer_field_proxy_type
 
         implicit none
 
-        class(extract_PSyDataType), intent(inout), target :: this
+        class(extract_PsyDataType), intent(inout), target :: this
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/standalone/lfric/kernel_data_standalone.jinja]: Done.

File: lib/nan_test/lfric/nan_test.jinja [hunk 31]
```diff
@@ -1,7 +1,14 @@
+{# Added this as jinja code so that it is understood that the
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/nan_test/lfric/nan_test.jinja]: Done ... here and everywhere.

File: lib/nan_test/lfric/nan_test.jinja [hunk 32]
```diff
@@ -1,7 +1,14 @@
...
+   comment does not apply to THIS file. #}
+{{ "! ================================================== !" }}
+{{ "! THIS FILE IS CREATED FROM THE JINJA TEMPLATE FILE  !" }}
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/nan_test/lfric/nan_test.jinja]: Done ... here and everywhere.

File: lib/nan_test/lfric/nan_test.jinja [hunk 33]
```diff
@@ -1,7 +1,14 @@
...
+{{ "! DO NOT MODIFY DIRECTLY                             !" }}
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/nan_test/lfric/nan_test.jinja]: Done ... here and everywhere.

File: lib/nan_test/lfric/nan_test.jinja [hunk 34]
```diff
@@ -53,93 +65,152 @@ module nan_test_psy_data_mod
...
+        character(*), intent(in)                           :: name
+        type(field_r{{prec}}_type), intent(in)                   :: value
 
-    end subroutine DeclareField
+    end subroutine DeclareField_r{{prec}}
 
     ! -------------------------------------------------------------------------
-    !> This subroutine declares LFRic integer-valued fields. No functionality
-    !! is needed for NAN checking, so it is just an empty function.
+    !> This subroutine checks whether a real-valued LFRic field has NAN or
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/nan_test/lfric/nan_test.jinja]: Done ... here and everywhere.

File: lib/nan_test/lfric/nan_test.jinja [hunk 35]
```diff
@@ -53,93 +65,152 @@ module nan_test_psy_data_mod
...
-        type(integer_field_type), intent(in) :: value
+        character(*), intent(in)                           :: name
+        type(field_r{{prec}}_type), intent(in)                   :: value
 
-    end subroutine DeclareIntField
+        type(field_r{{prec}}_proxy_type)                         :: value_proxy
+
+        if (.not. is_enabled) return
+
+        if (this%verbosity>1) then
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/nan_test/lfric/nan_test.jinja]: Done.

File: lib/nan_test/lfric/nan_test.jinja [hunk 36]
```diff
@@ -53,93 +65,152 @@ module nan_test_psy_data_mod
...
+        endif
+        value_proxy = value%get_proxy()
+        call this%ProvideVariable(name, value_proxy%data)
+
+    end subroutine ProvideField_r{{prec}}
 
     ! -------------------------------------------------------------------------
-    !> This subroutine checks whether a real-valued LFRic field has NAN or
-    !! infinite floating-point values.
+    !> This subroutine declares LFRic real-valued field vectors. No
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/nan_test/lfric/nan_test.jinja]: Done ... here and everywhere.

File: lib/nan_test/lfric/nan_test.jinja [hunk 37]
```diff
@@ -53,93 +65,152 @@ module nan_test_psy_data_mod
...
+            call this%ProvideVariable(name//trim(index_string), value(i))
+        enddo
+
+    end subroutine ProvideFieldVector_r{{prec}}
+
+    {% endfor %}
 
-    end subroutine ProvideField
+    ! -------------------------------------------------------------------------
+    !> This subroutine declares LFRic integer-valued fields. No functionality
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/nan_test/lfric/nan_test.jinja]: Done ... here and everywhere.

File: lib/read_only/lfric/read_only.jinja [hunk 38]
```diff
@@ -1,7 +1,14 @@
+{# Added this as jinja code so that it is understood that the
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/read_only/lfric/read_only.jinja]: Done ... here and everywhere.

File: lib/read_only/lfric/read_only.jinja [hunk 39]
```diff
@@ -1,7 +1,14 @@
...
+   comment does not apply to THIS file. #}
+{{ "! ================================================== !" }}
+{{ "! THIS FILE IS CREATED FROM THE JINJA TEMPLATE FILE  !" }}
+{{ "! DO NOT MODIFY DIRECTLY                             !" }}
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/read_only/lfric/read_only.jinja]: Done ... here and everywhere.

File: lib/read_only/lfric/read_only.jinja [hunk 40]
```diff
@@ -58,71 +71,66 @@ module read_only_verify_psy_data_mod
...
                            ProvideIntFieldVector
 
     end type read_only_verify_PSyDataType
 
 contains
 
+    {% for prec in ALL_PREC %}
     ! -------------------------------------------------------------------------
-    !> This subroutine declares a real-valued (double precision) field based on
+    !> This subroutine declares a real-valued {{prec}} bit field based on
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/read_only/lfric/read_only.jinja]: Done ... here and everywhere.

File: lib/read_only/lfric/read_only.jinja [hunk 41]
```diff
@@ -168,7 +175,79 @@ subroutine ProvideField(this, name, value)
 
         this%next_var_index = this%next_var_index + 1
 
-    end subroutine ProvideField
+    end subroutine ProvideField_r{{prec}}
+
+    ! -------------------------------------------------------------------------
+    !> This subroutine declares a vector of real-valued {{prec}}-bit
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/read_only/lfric/read_only.jinja]: Done ... here and everywhere.

File: lib/read_only/lfric/read_only.jinja [hunk 42]
```diff
@@ -168,7 +175,79 @@ subroutine ProvideField(this, name, value)
...
+            write(index_string, '("(",i0,")")') i
+            call this%ProvideField_r{{prec}}(name//trim(index_string), value(i))
+        enddo
+
+    end subroutine ProvideFieldVector_r{{prec}}
+
+    {% endfor %}
+
+    ! -------------------------------------------------------------------------
+    !> This subroutine declares an integer field based on the LFRic
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/read_only/lfric/read_only.jinja]: Done ... here and everywhere.

**Comment by hiker** [comment]
Thank you very much for the very detailed review (feel free next time to just post one comment and tell me to fix it everywhere :)  ).

I have fixed most (all?) of the suggestions in all jinja files, even the ones not originally included in the PR. In hindsight maybe that's a bit confusing, but at least now all .jinja files under lib are consistent. I would suggest not to review the new files that have only changes in comments (1d -> 1D; "." and "!" in warning message,; jinja->Jinja, !>@brief). I've commented them all as "Done ... here and everywhere." :)

Sorry for the compilation problem, I somehow forgot to update the dependency file :( 

Re 'make clean' not removing the library: that's on purpose, there is a special target 'allclean' which removes all libraries external to the example.

Should be ready for next review.

**Pull request merged**